src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,118 @@
+package org.folio.circulation.domain.validation;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+import static org.folio.circulation.support.Result.ofAsync;
+import static org.folio.circulation.support.Result.succeeded;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.apache.commons.lang3.StringUtils;
+
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.policy.LoanPolicyRepository;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ValidationErrorFailure;
+
+public class ItemLimitValidator {
+  private final Function<String, ValidationErrorFailure> itemLimitErrorFunction;
+  private final LoanRepository loanRepository;
+  private final LoanPolicyRepository loanPolicyRepository;
+
+  public ItemLimitValidator(Function<String, ValidationErrorFailure> itemLimitErrorFunction,
+    LoanRepository loanRepository, LoanPolicyRepository loanPolicyRepository) {
+
+    this.itemLimitErrorFunction = itemLimitErrorFunction;
+    this.loanRepository = loanRepository;
+    this.loanPolicyRepository = loanPolicyRepository;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> refuseWhenItemLimitIsReached(
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return ofAsync(() -> loan.getLoanPolicy().getRuleConditions())
+      .thenComposeAsync(result -> result.failAfter(ruleConditions -> isLimitReached(ruleConditions, records),
+        ruleConditions -> {
+          String message = getErrorMessage(ruleConditions);
+          return itemLimitErrorFunction.apply(String.format("Patron has reached maximum item limit of %d items %s",
+            itemLimit, message));
+        }))
+      .thenApply(result -> result.map(v -> records));
+  }
+
+  private CompletableFuture<Result<Boolean>> isLimitReached(List<String> ruleConditions, LoanAndRelatedRecords records) {
+
+    if (!isRuleMaterialTypePresent(ruleConditions) && !isRuleLoanTypePresent(ruleConditions)) {
+      return ofAsync(() -> false);
+    }
+
+    Item item = records.getLoan().getItem();
+    String materialType = item.getMaterialType() != null
+      ? item.getMaterialType().getString("name")
+      : null;
+    String loanType = item.getLoanTypeName();
```

Comment:
Why are the named used here instead of the IDs?

Reply:
Thanks, updated, 
now id is used.

Reply:
Does that mean that the material type and loan type records no longer have to be fetched for this validation to be applied?

Reply:
Hm... I'm not sure... without loan type and material type fetching we will get loan.getItem() == null in the code below: 
`expectedMaterialTypeId.equals(loan.getItem().getMaterialType().getString("id"));`
We still need to compare material type and loan type between item and DB loan record.
Or correct me if I'm wrong please...

Reply:
I don’t understand how not fetching the material type means the item is null. Please could you help me understand better.

Reply:
I mean that in LoanRepository#findOpenLoansByUserIdWithItem I fetch loans with items. Item contains material type and loan type. 
We use them material type and loan type in ItemLimitValidator.  If we don't fetch item for loan we will not be able to get material type and loan type, because loan doesn't have material type or loan type ids...
Here is the screenshot in debug: 
![image](https://user-images.githubusercontent.com/53909129/71719414-f9096980-2e26-11ea-98bb-bc3d328b5b5e.png)

Reply:
I'm not suggesting not fetching the item.  I'm suggesting that the item can be fetched without fetching the loan type or material type.

`findOpenLoansByUserIdWithItem` was added in this change. Was loading the material type and loan type included because they were needed by the validator? 

If they are no longer needed by the validator, could an existing method for the loans repository be used that does not load them?

Reply:
Yes, material type and loan type included because they are still needed by the validator

Reply:
Why are they still needed by the validator? If it is using the IDs the referenced records are not needed. 

The material type and loan type from the rules can be obtained from the original item.

What am I missing?

Reply:
@marcjohnson-kint , yes, you are right. 
I have removed material type and loan type fetching and updated the validator. Thanks!